sql/2021/cdn/cdn_usage_by_site_rank.sql

```diff
@@ -0,0 +1,21 @@
+#standardSQL
+# cdn_usage_by_site_rank.sql : Distribution of HTML pages served by CDN vs Origin by rank
+
+SELECT client, cdn, rank,
+  COUNT(0) AS num_requests,
+  SUM(COUNT(0)) OVER (PARTITION BY client, cdn, rank) AS total_compressed,
+  COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client, cdn, rank) AS pct
+FROM (
+  SELECT
+    client,
+    IF(IFNULL(NULLIF(REGEXP_EXTRACT(_cdn_provider, r'^([^,]*).*'), ''), 'ORIGIN') = "ORIGIN", "ORIGIN", "CDN") AS cdn,
+    rank
+  FROM
+    `httparchive.almanac.requests`
+  WHERE
+    date = '2021-07-01' AND
+    firstHtml AND
+    rank IS NOT NULL
+  )
+GROUP BY client, cdn, rank
+ORDER BY client, rank, cdn
```

Comment:
I don't understand this query @paulcalvano 

`COUNT(0)` and `SUM(COUNT(0)) OVER (PARTITION BY client, cdn, rank)` will be the same since you use: `GROUP BY client, cdn, rank`?

Which also means the pct is always 100%

Thought I was going mad, but ran it and that's what BigQuery thinks too!

Also this is not the way `rank` works, as need to include each lower rank to get "Top 1,000@, "Top 10,000"...etc. rather than "Top 1,000", "Next 9,000", "Next 90,000"...etc.
See: https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Analysts'-Guide#rank

Reply:
I updated it to remove these clauses.    I didn't run with them when I created the results sheet for this, so not sure how it snuck into the version I checked in :) 

```
  SUM(COUNT(0)) OVER (PARTITION BY client, cdn, rank) AS total_compressed,
  COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client, cdn, rank) AS pct
```

Also, I'm very confused by what you mean about rank not working that way.  The way it is written in your example seems to double count some sites by rolling the lower rank groups up into the larger rank groups.

Reply:
Yes that's intentional.

The top 10,000 sites also includes the top 1,000 sites.
So we are converting from distinct "bucketed" groups to "top" groups, which are inclusive of lower groups.

We think most readers will expect it to work that way, rather than in distinct buckets. It's how other lists (e.g. Alexa ranking lists) are done. Most people wouldn't expect the Alexa top million to only have 900,000 sites because you have to merge it with the Alexa top 100,000 sites too (and beyond).